---
title: "Formatting, exploring and imputing genomic data"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  bookdown::html_document2:
    toc: true
    toc_depth: 4
    number_sections: true
bibliography: references.bib
editor_options: 
  chunk_output_type: console
---

<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>

<style type="text/css">
div.main-container {
  max-width: 2000px;
  margin-left: auto;
  margin-right: auto;
}
</style>


```{css, echo=FALSE}
pre {
  max-height: 150px;
  overflow-y: auto;
  overflow-x: auto;
  font-size: 10px;
}

body{
  font-size: 12px;
}
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width = 5,fig.height = 4,cache=F)
options(width = 300)
library(knitr)
library(dplyr)
library(tidyverse)
library(readxl)
library(magrittr)
library(kableExtra)
library(genetics)
library(xtable)
library(reshape2)
library(janitor)
library(diffdf)
library(parallel)
library(here)
```


```{r Functions}
# Function to build tables
kable_mydf <- function(x, boldfirstcolumn,font_size){
  x %>% 
  kable() %>%  
  kable_styling(font_size=font_size,
                bootstrap_options = c("stripped","hover", "condensed"), full_width = F) %>% 
  {if(boldfirstcolumn == TRUE) column_spec(., 1, bold = T) else .}
}
```


# Raw genomic data

## Loading

We load the raw genomic data available in the DRYAD repository. 

A subset of rows are shown below (rows 40 to 50 and columns 1 to 13).

```{r LoadRawGenomicData}
# Load the raw genomic data:
data <- read.csv(here("data/DryadRepo/RawGenomicData.csv"))

# viz
data[40:50,1:13] %>% kable_mydf(boldfirstcolumn = F, font_size = 11)
```

Below is a subset of the whole genomic dataset. NAs are indicated with `---`.

Meaning of the columns:

  - `clone`: clone ID.
  
  - `assay`: Assay in which the clone was genotyped, either the Infinium assay (`only_Inf`), the Axiom assay (`only_Affx`) or both assays (`both_Inf_Affx`).
  
  - `snp_1` --> `snp_14016`:  genotype for each of the 14,016 SNPs.



## Formatting

For each SNP, we convert the letters (e.g. A/A, A/G) to numbers (0,1 or 2).

```{r FormattingRawGenomicDataFromLettersToNumbers}
# We apply a function to each column (i.e SNP) of the raw genomic data, 
# after removing the first two columns, i.e. the clone ID and the assay in which the clone was genotyped

list_alleles <- lapply(colnames(data[,-c(1,2)]), function(x){ 
  
  vec <- data[,x]
  
  tab <- table(data[,x]) %>%  as.matrix() %>%  t() %>% as.data.frame() %>% dplyr::select(-contains("---"))
  
  
  if(length(colnames(tab))==1){         # monomorphic case, e.g. only A/A
    major <- names(which.max(tab))
    
    case_when(vec == major ~ 0) 
    
  } else if(length(colnames(tab))==2){  # case where there is no homozygote with the rare allele, e.g. A/A and A/G
    
    major <- names(which.max(tab))
    mid <- names(which.min(tab))
    
    case_when(vec == major ~ 0,
              vec == mid   ~ 1) 
    
  } else if (length(colnames(tab))==3){ # common case, e.g. A/A, A/G and G/G
    
    if(tab[,1]==tab[,3]){               # to account for cases where there is the same number of A/A and G/G 
      major <- colnames(tab)[1]
      minor <- colnames(tab)[3]
      } else {
    major <- names(which.max(tab[,c(1,3)]))
    minor <- names(which.min(tab[,c(1,3)]))
      }
    mid <- setdiff(names(tab),c(major,minor))  
    
    case_when(vec == major ~ 0,
              vec == minor ~ 2,
              vec == mid   ~ 1) 
  }
})

names(list_alleles) <- colnames(data[,-c(1,2)]) # re-attribute the SNP ID to list names

df <- list_alleles %>% 
  as_tibble() %>% 
  t()

colnames(df) <- data$clone # attribute clone ID

# viz
df[1:11,40:50] %>% kable_mydf(boldfirstcolumn = F, font_size = 11)
```



# Filtering genomic data

## MAC and monomorphic SNPs{#MAC}

```{r IdentifyingMonomorphicSNPs}
sum_minor_allele <- df %>% 
  t() %>% 
  as_tibble() %>% 
  sapply(function(x) sum(x,na.rm=T)) %>% 
  as.data.frame() %>%
  set_colnames(c("sum")) %>% 
  rownames_to_column(var = "snp") %>% 
  as_tibble() 

mono <- sum_minor_allele %>% filter(sum==0)
low_mac <- sum_minor_allele %>% filter(sum %in% c(1,2))
```

`r nrow(mono)` SNPs are monomorphic and therefore, we are going to remove them.


`r nrow(low_mac)` SNPs have one or two minor allele count. We may want to remove these SNPs two as these counts may be due to PCR or sequencing errors. We will see after removing the MAF if these SNPs have also been removed. 


```{r RemovingMonomorphicSNPs}
df <- df[!(row.names(df) %in% mono$snp),]
```

<span style="color: red;">There are `r nrow(df)` non-monomorphic SNPs left.</span>


## Removing the ROD population

We remove the clones from the ROD (Rodoiros) population, as they come from a plantation of material from Central Spain in the Iberian Atlantic region.

```{r RmROD}
df <- df %>% 
  as.data.frame() %>% 
  dplyr::select(-contains("ROD"))
```


## Filtering based on NAs and MAF

### Missing data per clone

```{r MissingDataPerClone, warning=F, fig.height=4,fig.width=6}
tab <- df %>% 
  as_tibble() %>% 
  sapply(function(x) sum(is.na(x))*100/nrow(df)) %>% 
  as.data.frame() %>%
  set_colnames(c("na_freq")) %>% 
  rownames_to_column(var = "clone") %>% 
  as_tibble() %>% 
  arrange(na_freq) %>% 
  mutate(id=1:nrow(.))

plot(tab$id,tab$na_freq,
     pch=20,
     xlab="Clone number",
     ylab="Percentage of missing data per clone")
```

`r nrow(tab[tab$na_freq<20,])` clones have less than 20% missing data, `r nrow(tab[tab$na_freq<18,])` clones have less than 18% and `r nrow(tab[tab$na_freq<15,])` have less than 15%.

```{r RemoveCloneMissingData}
# Specifying the chosen threshold:
threshold_NA_per_clone <- 18

# Removing clones with more than 18% missing data
df <- df %>% dplyr::select(tab[tab$na_freq<threshold_NA_per_clone,]$clone)
```

**<span style="color: red;">We keep the `r nrow(tab[tab$na_freq<threshold_NA_per_clone,])` clones that have less than `r threshold_NA_per_clone`% missing data.</span>**

### Minor allele frequencies

We want to remove *minor allele frequencies* (MAF) for gene-environment association analyses, but not for population genetic structure estimation. Therefore, we produce two datasets: one with MAF and the other without, that we store in a list.

```{r MAF, warning=F, fig.height=4,fig.width=6}
tab <- df %>% 
  t() %>% 
  as_tibble() %>% 
  sapply(function(x) sum(x,na.rm=T)*100/(2*nrow(.))) %>% 
  as.data.frame() %>%
  set_colnames(c("MAF")) %>% 
  rownames_to_column(var = "snp") %>% 
  as_tibble() %>% 
  arrange(MAF) %>% 
  mutate(id=1:nrow(.))

plot(tab$id,tab$MAF,
     pch=20,
     xlab="SNP number",
     ylab="Minor allele frequencies")
```

```{r RmMissingDataMAF}
list_geno <- list(list_withmaf=list(df = df),
                  list_withoutmaf=list(df = df[!(row.names(df) %in% tab[tab$MAF<1,]$snp),]))
```

**<span style="color: red;">`r nrow(tab[tab$MAF<1,])` SNPs have MAF < 1%. In the dataset without MAF, we remove the SNPs with MAF < 1%: there are `r nrow(list_geno$list_withoutmaf$df)` SNPs left.</span>** 


### Missing data per SNP

```{r MissingDataPerSNP, warning=F}
remove_NA_perSNP <- function(x){
  
tab <- x$df %>% 
  t() %>% 
  as_tibble() %>% 
  sapply(function(x) sum(is.na(x))*100/nrow(.)) %>% 
  as.data.frame() %>%
  set_colnames(c("na_freq")) %>% 
  rownames_to_column(var = "snp") %>% 
  as_tibble() %>% 
  arrange(na_freq) %>% 
  mutate(id=1:nrow(.))

df <- x$df[(row.names(x$df) %in% tab[tab$na_freq<20,]$snp),] %>% 
  dplyr::select(order(colnames(x$df)))

list(df=df,tab=tab)
}

list_geno <- lapply(list_geno,remove_NA_perSNP)
```

In the dataset without MAF:

`r nrow(list_geno$list_withoutmaf$tab[list_geno$list_withoutmaf$tab$na_freq<20,])` SNPs have less than 20% missing data.
`r nrow(list_geno$list_withoutmaf$tab[list_geno$list_withoutmaf$tab$na_freq<15,])` SNPs have less than 15% missing data.
`r nrow(list_geno$list_withoutmaf$tab[list_geno$list_withoutmaf$tab$na_freq<10,])` SNPs have less than 10% missing data.
`r nrow(list_geno$list_withoutmaf$tab[list_geno$list_withoutmaf$tab$na_freq<5,])` SNPs have less than 5% missing data.
`r nrow(list_geno$list_withoutmaf$tab[list_geno$list_withoutmaf$tab$na_freq<1,])` SNPs have less than 1% missing data.

**<span style="color: red;">We keep SNPs with less than 20% missing data and so there are `r nrow(list_geno$list_withoutmaf$df)` SNPs left in the dataset without MAF and `r nrow(list_geno$list_withmaf$df)` in the dataset with MAF.</span>**


```{r PlotMissingDataPerSNP, fig.height=4,fig.width=6}
plot(list_geno$list_withoutmaf$tab$id,list_geno$list_withoutmaf$tab$na_freq,
     pch=20,
     xlab="SNP number",
     ylab="Percentage of missing data per SNP",
     main="Genomic dataset without MAF")

# Add horizontal lines at 20% and 15%
abline(h = 20, col = "red", lty = 2)  # Red dashed line at 20%
```



## Filtering correlated SNPs

In this section, we estimate the statistical correlations among SNPs and examine whether these correlations can be explained by the closed locations of SNPs on the genome or not. 

To do this, we use the `genome_position.xlsx` dataset, which contains information on SNP positions on the genome. A cleaned-up version of this dataset available in the DRYAD repository; see section \@ref(ListSNPs).
 
### Statistically correlated SNPs

We extract the pairs of SNPs with a correlation coefficient of 1 in the dataset not filtered for MAF.

```{r GenerateFilePairsCorrelatedSNPs, warning=F, eval=F}
# Invert the genomic matrix
tdf <- list_geno$list_withmaf$df %>% t() %>% as.data.frame() 

# Create a matrix of correlation for all SNPs
matcor <- cor(tdf, use = "pairwise.complete.obs")

# Keep only the upper triangle of the matrix and remove values in the diagonal
matcor[lower.tri(matcor,diag=T)] <- NA

# Arrange the pairs of correlated SNPs in two columns
snpcor <- matcor %>% 
  melt() %>% 
  na.omit() %>% 
  dplyr::filter(value==1) %>% 
  dplyr::select(-value) %>% 
  set_colnames(c("SNP1","SNP2"))

rm(matcor)

snp_match_1 <- read.csv(here("data/GenomicData/SnpCodesMatching.csv")) %>% 
  set_colnames(paste0(colnames(.),"_SNP1")) %>% 
  rename(SNP1 = snp_ID_SNP1)

snp_match_2 <- read.csv(here("data/GenomicData/SnpCodesMatching.csv")) %>% 
  set_colnames(paste0(colnames(.),"_SNP2")) %>% 
  rename(SNP2 = snp_ID_SNP2)
  
snpcor <- snpcor %>% 
  left_join(snp_match_1, by = "SNP1") %>% 
  left_join(snp_match_2, by = "SNP2")

# save file (file sent to Santi 14/02/2023)
# snpcor %>% write_csv(here("data/GenomicData/CorrelatedSNPs.csv"),append=FALSE) 
```

```{r LoadCorrelatedSNPsDataset}
# we load the dataset of pairs of SNPs with a correlation coeff of 1 that I sent to Santi the 14/02/2023
snpcor <- read.csv(here("data/GenomicData/CorrelatedSNPs.csv"))
```


### SNP positions on the genome

This part aims at understanding how the information about SNP position on the genome is stored in the excel file `genome_position.xlsx` (sent by Santi the 14/02/2023). In this excel file, the sheet `base_table` provides the information about the SNP positions on the genome. The position of the SNPs on the genome is given by the columns `scaffold` and `genome_position`. However, it is important to note that some SNPs have **alternative names**, referring to **different positions** on the genome. The alternative names are given in the column `annotation`. 

```{r SNPpositionInformationSheetBaseTable}
# we load the dataset with information about the SNP position on the genome
# -------------------------------------------------------------------------

snppos <- read_excel(here("data/GenomicData/genome_position.xlsx"), sheet = "base_table") %>% 
  dplyr::rename(snp_ID = Juliette_snp_ID,
                original_ID = Original_ID,
                affx_ID = affy_snp_id,
                infinium_ID = Infinium_ID,
                genome_position =`position (bp)`,
                annotation =`Annotation (including alternative names)`) %>% 
  dplyr::select(snp_ID, original_ID, affx_ID, infinium_ID, scaffold, genome_position, annotation)

snppos[1:5,] %>% kable_mydf(boldfirstcolumn = F, font_size = 11)

# Attributing position information to the two SNPs of each pair of correlated SNPs
# --------------------------------------------------------------------------------

snppos_SNP1 <- snppos %>% 
  set_colnames(paste0(colnames(.),"_SNP1")) %>% 
  dplyr::rename(SNP1 = snp_ID_SNP1)
  
snppos_SNP2 <- snppos %>% 
  set_colnames(paste0(colnames(.),"_SNP2")) %>% 
  dplyr::rename(SNP2 = snp_ID_SNP2)

snpcor <- snpcor %>% 
  inner_join(snppos_SNP1, by = c("SNP1", "original_ID_SNP1", "affx_ID_SNP1", "infinium_ID_SNP1")) %>% 
  inner_join(snppos_SNP2, by = c("SNP2", "original_ID_SNP2", "affx_ID_SNP2", "infinium_ID_SNP2")) %>% 
  plyr::arrange(SNP1,SNP2) %>% 
  as_tibble()
```

Then, let's look at the sheet `correlated_SNPs` of the excel file `genome_position.xlsx`. In this sheet, there is the information of the position on the genome for each of the SNPs of the correlated SNPs pairs. The column `ToDo` corresponds to Santi's suggestion about whether we should keep or not the two correlated SNPs:

  - `keep_both` when the two SNPs are located on different scaffolds.
  
  - `keep_one` when the two SNPs are located on the same scaffold.

```{r CorrelatedSNPsToRemoveSanti, message=F}
sheet2 <- read_excel(here("data/GenomicData/genome_position.xlsx"), sheet = "correlated_SNPs") %>% 
  dplyr::select(contains("SNP1"),contains("SNP2"), contains ("scaffold"), contains("position"), contains("annotation"), ToDo) %>% 
  dplyr::rename(scaffold_SNP1 = "scaffold...5",
                scaffold_SNP2 = "scaffold...13",
                genome_position_SNP1 = "position (bp)...6",
                genome_position_SNP2 = "position (bp)...14",
                annotation_SNP1 = "annotation (including alternative names)...7",
                annotation_SNP2 = "annotation (including alternative names)...15") %>% 
  plyr::arrange(SNP1,SNP2) %>%
  mutate(across(where(is.character),~na_if(.,"NA")))
```


In the sheet `base_table`,`r snpcor %>% filter(scaffold_SNP1 == scaffold_SNP2) %>% nrow()` pairs of correlated SNPs are located on the same scaffold according to the column `scaffold`.

In the sheet `correlated_SNPs`, Santi suggested to keep only one SNPs for `r sheet2 %>%  dplyr::filter(ToDo=="keep_one") %>% nrow()` pairs of correlated SNPs.

So, for `r sheet2 %>%  dplyr::filter(ToDo=="keep_one") %>% nrow() - snpcor %>% filter(scaffold_SNP1 == scaffold_SNP2) %>% nrow()` pairs of SNPs, Santi suggested to keep only one SNPs whereas those SNPs do not have the same values in the `scaffold` column. These differences are explained by the fact that some SNPs have alternative names, and so if we account for these alternative names, the SNPs are indeed located on the same scaffold.

For two SNPs, different SNP names were used in the two sheets to fill the `scaffold` and `genome_position` columns. But if we look at the alternative names of those SNPs, there are located on the same scaffolds.

```{r ComparingTwoDatasets, message=F, warning=F}
diffdf(sheet2 %>% dplyr::select(-ToDo) , snpcor)

# Information in the sheet `correlated_SNPs`
# ------------------------------------------

anti_join(sheet2 %>% dplyr::select(-ToDo) , snpcor) %>% 
  dplyr::select(SNP1, SNP2, scaffold_SNP1, genome_position_SNP1, annotation_SNP1) %>% 
  kable_mydf(boldfirstcolumn = F, font_size = 11)


# Information in the sheet `base_table`
# -------------------------------------

anti_join(snpcor, sheet2 %>% dplyr::select(-ToDo) ) %>% 
  dplyr::select(SNP1, SNP2, scaffold_SNP1, genome_position_SNP1, annotation_SNP1) %>% 
  kable_mydf(boldfirstcolumn = F, font_size = 11)
```

For the 12 other pairs of SNPs, the columns `scaffold` are not the same for the two SNPs of the pair. But if we look at their alternative names, the two SNPs are in fact on the same scaffold, and this is why Santi indicated to keep only one of the two SNPs (`keep_one` in the column `ToDo`). 

```{r IdentifyCorrelatedSNPsLocatedSameScaffold, message=F}
diff <- sheet2 %>% 
  dplyr::filter(ToDo=="keep_one") %>% 
  inner_join(snpcor) %>% 
  filter(scaffold_SNP1 != scaffold_SNP2)
```

So, all is ok, in the next section, we will use the information in the sheet `correlated_SNPs` to filter the SNPs that have a correlation coefficient of 1 with another SNPs located on the same scaffold.


### To prune or not to prune?

Initially, for the pairs of SNPs located on the same scaffold and with a correlation coefficient of 1 (so the same genotypes across all the individuals), we wanted to remove one of the two SNPs of the pair. The idea behind was that those SNPs may be duplicates. However, after some discussion with Thibaut Capblancq, we decided not to prune. Some explanations below.

**First question: should we prune our genomic dataset?**

*Arguments not to prune*: it is important to have all the signal possible and we do not expect that there will be a lot of variation in terms of the number of SNPs in genomic regions with different signals.

*Arguments to prune*: some SNP-rich regions that are very physically linked (= in physical proximity on the genome) will drive the signal while they are not more important than others, by pruning we give the same weight to all sampled genomic regions, regardless of the SNP-richness of these regions.

**Second question: if we decide to prune, how do we prune?**

In general, we only prune according to the **physical distance** of the SNPs along the genome and **not their statistical association** (ie similarity of genotypes). Normally people estimate the average linkage disequilibrium along the genome and decide on a distance (in bp) below which they will keep only one SNP (at random). For instance, for maritime pine, we lose half of the linkage disequilibrium in about **200bp** so we may want to prune the dataset to keep only **one SNP every 500 bases** (this is what Thibaut Capblancq does with red spruce). 

In our case, we have the information about the **SNP scaffold**. A scaffold is a non-contiguous assembly of genomic sequences -contigs- separated by gaps of known length and is about 1000 bp in length. So we may want to keep one or two SNPs per scaffold.


<span style="color: red;">We therefore decided not to prune the dataset, as we were concerned about removing a large part of the signal by keeping only one SNP per 500 bp.</span>

However, we decided to:

  - check whether candidate SNPs include some SNPs located on the same scaffold. If yes, we may prune them at this step (keep only one when different candidate SNPs are located on the same scaffold).
  
  - redo some parts of the analyses (eg calculating the genomic offset with the best performing method) with a pruned dataset.
  
  
### Filtering correlated SNPs on the same scaffold

The script in this section are not evaluated as we decided not the prune the dataset based on the statistical relationships of the SNPs. 

```{r ExtractCorrelatedSNPsToRemove, message=F}
# We extract the SNPs that we are going to remove
corrSNPs_to_remove <- read_excel(here("data/GenomicData/genome_position.xlsx"), sheet = "correlated_SNPs") %>% 
  dplyr::select(SNP2, ToDo) %>% 
  dplyr::filter(ToDo=="keep_one") %>%  # select only the cases in which the two SNPs are one the same scaffold
  pull(SNP2) %>% # SNPs in the SNP2 column are those that we are going to rm
  unique()
```

If we choose to remove SNPs on the same scaffold and with the same genotypes across all individuals, we would remove **`r length(corrSNPs_to_remove)` SNPs.**

```{r RemoveCorrelatedSNPs, eval=F}
for(i in 1:2) list_geno[[i]]$df <- list_geno[[i]]$df %>% dplyr::filter(!(rownames(.) %in% corrSNPs_to_remove)) 
```

<span style="color: red;">In the genomic dataset with MAF, **`r list_geno$list_withmaf$df %>% nrow()` SNPs left**.
In the genomic dataset without MAF, **`r list_geno$list_withoutmaf$df %>% nrow()` SNPs left**.</span>

## Filtering minor allele counts (MAC)

We check whether the SNPs with very low minor allele counts (MAC < 3) identified in section \@ref(MAC) have been removed from the two genomic datasets after the filtering steps. 

```{r CheckingRemainingMACnumber}
nb_mac_geno_with_maf <- rownames(list_geno$list_withmaf$df)[rownames(list_geno$list_withmaf$df) %in% low_mac$snp] %>% 
  length()
nb_mac_geno_without_maf <- rownames(list_geno$list_withoutmaf$df)[rownames(list_geno$list_withoutmaf$df) %in% low_mac$snp] %>%
  length()

tibble("Genomic dataset" = c("With MAF","Without MAF"),
       "Number of remaining SNPs with MAC < 3" = c(nb_mac_geno_with_maf, nb_mac_geno_without_maf)) %>% 
  kable_mydf(boldfirstcolumn = F, font_size = 11)
```

**We remove the SNPs with MAC < 3 from the dataset with MAF.**

```{r RemoveRemainingMAC}
list_geno$list_withmaf$df <- list_geno$list_withmaf$df %>% dplyr::filter(!(rownames(.) %in% low_mac$snp)) 
```

There are `r nrow(list_geno$list_withmaf$df)` left in the genomic dataset with MAF.


# Exporting the formatted and filtered genomic datasets

##  Allele counts (i.e. 0/1/2)

We export the dataset containing allele counts (coded as 0, 1 or 2) in the DRYAD repository. SNPs in rows and genotypes in columns.

  - column 1: `snp_ID`: SNP ID.
  
  - columns 2 to 455: `ALT10` to `VER9`:  Clone (i.e., genotype) ID.

```{r ExportingFormattedGenomicDataAlleleCounts}
for(i in 1:length(list_geno)){
  specMAF <- names(list_geno[i])[1] %>% str_sub(6, -1) #%>% str_to_upper()
  
  list_geno[[i]]$df %>% 
  rownames_to_column(var="snp_ID") %>% 
  write_csv(here(paste0("data/DryadRepo/FormattedFilteredGenomicData_AlleleCounts_",specMAF,".csv")),
            append=FALSE)
}

# Reminder
# to download the dataset with snp_ID as first column: use `read.csv` with missing row.names
# to download the dataset with snp_ID as row names: use `read.csv` with row.names = 1
```

`r sum(is.na(list_geno$list_withmaf$df))` missing data in the dataset with MAF, and `r sum(is.na(list_geno$list_withoutmaf$df))` missing data in the dataset without MAF.

##  Allele frequencies

We calculate allele frequencies at the population level for the dataset filtered for MAF. And we export the allle frequencies in the DRYAD repository. SNPs in columns and populations in rows.

  - column 1: `pop`: population.
  
  - columns 2 to 9818: SNP ID.
  
```{r ExportingFormattedGenomicDataAllelefrequencies}
# Only useful for the dataset without MAF

compute_allele_freq <- function(x){
  
x %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var="pop") %>% 
  mutate(pop=str_sub(pop,1,3)) %>% 
  group_by(pop) %>% 
  summarise_all(~sum(., na.rm = TRUE)/((n()-sum(is.na(.)))*2))
  
}

df_freq <- compute_allele_freq(list_geno$list_withoutmaf$df) 

df_freq %>% 
  write_csv(here(paste0("data/DryadRepo/FormattedFilteredGenomicData_AlleleFrequencies_withoutmaf.csv")),
            append=FALSE)

df_freq[1:10,1:8] %>% 
  mutate(across(where(is.numeric), round, 2)) %>% 
  kable_mydf(boldfirstcolumn = T, font_size = 12)
```

`r sum(is.na(df_freq))` missing data in this dataset.


## List of SNPs{#ListSNPs}

We create a dataset with information on the genome position and the different SNP IDs after the filtering steps (but including the SNPs with minor allele frequencies). This dataset will be updated with information on outlier SNPs identified in the GEAs and will then be exported in the DRYAD repository. SNPs are in rows.

Meaning of the columns:

  1. `snp_ID`: SNP ID used in the present study.
  2. `original_ID`: Original ID of the SNP.
  3. `affx_ID`: SNP ID in the Axiom assay.
  4. `infinium_ID`: SNP ID in the Infinium assay.
  5. `scaffold/contig`: contig on which the SNP is located. The term scaffold is also used as some SNPs were obtained from the alignment of NGS short-reads from a pseudoreference genome in *Pinus pinaster* that were called scaffolds (even though there are not really scaffolds).
  6. `genome_position`: position of the SNP on the scaffold/contig (in bp).
  7. `annotation`: SNP name (including possible alternative names).
  8. `MAF_filtering`: `REMOVED` for SNPs with minor allele frequencies, which are removed for GEAs analyses but not for estimating the neutral population genetic structure; `KEPT` otherwise.
  
Below are show the first 10 rows.
  
```{r ListSelectedSNPsWithGenomePositionInformation}
snps_withMAF_removed <- setdiff(rownames(list_geno$list_withmaf$df),rownames(list_geno$list_withoutmaf$df))

listsnps <- list_geno$list_withmaf$df %>% 
  rownames_to_column(var="snp_ID") %>%
  dplyr::select(snp_ID) %>% 
  inner_join(snppos, by="snp_ID") %>% 
  dplyr::rename("scaffold/contig"="scaffold") %>% 
  mutate(MAF_filtering=ifelse(snp_ID %in%snps_withMAF_removed, "REMOVED","KEPT")) %>% 
  write_csv2(here("data/GenomicData/SNPsInformation.csv"),append=FALSE)


listsnps[1:10,] %>% kable_mydf(font_size = 12, boldfirstcolumn = FALSE)
```


# Collinearity among SNPs

This part is only done on genomic data without MAF.

## Identifying duplicates

The first way to identify the duplicated SNPs is by using the function `duplicated`. 

```{r IdentifyPerfectDuplicates}
df <- read.csv(here("data/DryadRepo/FormattedFilteredGenomicData_AlleleCounts_withoutmaf.csv"), row.names = 1) 

dup <- df[duplicated(df) | duplicated(df, fromLast=TRUE), ]
```

There are `r nrow(dup)` duplicates (=SNPs that have exactly the same genotype as at least one other SNP).

Then, we determine how many copies there are for each duplicate. 

```{r CountCopiesDuplicates}
count_duplicates <- function(DF){
  x <- do.call('paste', c(DF, sep = '\r'))
  ox <- order(x)
  rl <- rle(x[ox])
  cbind(DF[ox[cumsum(rl$lengths)],,drop=FALSE],count = rl$lengths)
  
}
countdup <- count_duplicates(df %>% as.data.frame()) 

tab <- table(countdup$count) %>% 
  as.data.frame() %>% 
  setNames(c("SnpNumberWithinGroups","GroupNumber"))
  
tab  %>% kable_mydf(boldfirstcolumn = T, font_size = 12)
```

In this table, `r colnames(tab)[1]` is the number of duplicated SNPs within each group (ie groups of 2 SNPs, 3 SNPs, 4 SNPs, etc) and  `r colnames(tab)[2]` is the number of groups (or the number of non-duplicated SNPs for the first row). 

So it means that there are:

  - `r tab[1,2]` non-duplicated SNPs.
  - `r tab[2,2]` groups of `r tab[2,1]` duplicated SNPs.
  - `r tab[3,2]` groups of `r tab[3,1]` duplicated SNPs.
  - `r tab[4,2]` group of `r tab[4,1]` duplicated SNPs.
  - `r tab[5,2]` group of `r tab[5,1]` duplicated SNPs.
  - `r tab[6,2]` group of `r tab[6,1]` duplicated SNPs.
  - `r tab[7,2]` group of `r tab[7,1]` duplicated SNPs.

```{r ChecksNumberDuplicates, eval=F}
# checking that the numbers are coherent

# we should recover the total number of SNPs
nb_duplicates <- tab %>% 
  dplyr::mutate(product=as.numeric(SnpNumberWithinGroups)*as.numeric(GroupNumber)) %>% 
  dplyr::summarise(sum=sum(product))
nb_duplicates # that's ok!

# we should recover the total number of duplicated SNPs
nb_duplicates - tab$GroupNumber[1] # that's ok too!
```

But, there are two situations not accounted for by the function `duplicated`. 

  - First, if two SNPs have exactly the same genotype for all clones, except one clone where one of the two SNPs has a value of "1" for instance, and the other SNP has the value "NA". In this case, the two SNPs will not be considered as duplicates. 
  - Second, if the two SNPs covary perfectly but one have "0" when the other have "2", and vice versa, they will not be counted as duplicates neither. For instance:
      - SNP_1: 0 0 1 1 2 2 
      - SNP_2: 2 2 1 1 0 0 

To add these situations in the count of duplicated SNPs, I selected all the pairs of SNPs with a correlation coefficient of 1. 

```{r IdentifyCorrelatedSNPs, warning=F}
# Invert the genomic matrix
tdf <- df %>% t() %>% as.data.frame() 

# Create a matrix of correlation for all SNPs
matcor <- cor(tdf, use = "pairwise.complete.obs")
matcor[1:10,1:10] %>%  
  round(2) %>% 
  kable_mydf(boldfirstcolumn = T, font_size = 12)

# Keep only the upper triangle of the matrix and remove values in the diagonal
matcor[lower.tri(matcor,diag=T)] <- NA
matcor[1:10,1:10]  %>%  
  round(2) %>% 
  kable_mydf(boldfirstcolumn = T, font_size = 12)
```


<span style="color: red;">`r length(which(matcor==1))` pairs of SNPs have a correlation coefficient of 1.</span>

```{r ExtractPairsCorrelatedSNPs}
snpcor <- matcor %>% 
  melt() %>% 
  na.omit() %>% 
  dplyr::filter(value==1) %>% 
  dplyr::select(-value)

rm(matcor) # to free memory space
```


## Pairwise LD

 We calculate the *pairwise linkage disequilibrium* among SNPs with the `genetics` package.

```{r CalculateLD,eval=F}
# format genomic data for the `LD` function of the `genetics` package
LD <- data %>% 
  dplyr::select(-clone,-assay) %>% 
  dplyr::select(row.names(df)) %>% # keep only filtered SNPs
  # !!Comment!! do not merge this line with the previous one, it does not anymore if I do it
  makeGenotypes() 

# calculate LD (very very very long)
LD <- LD(LD)

saveRDS(LD, file=here("outputs/LD/LDGeneticsRpackage.rds"))
```

```{r IdentifyHighLDSNPs}
LDmat <- readRDS(here("outputs/LD/LDGeneticsRpackage.rds"))
LDmat <- LDmat$`R^2`

snpLD <- LDmat %>% 
  melt() %>% 
  na.omit() %>% 
  dplyr::filter(value>0.99) %>% 
  dplyr::select(-value)

rm(LDmat) # to free memory space
```

`r nrow(snpLD)` pairs of SNPs with LD > 0.99 (with LD estimated as the $r$ correlation coefficient among the SNPs).

```{r CommonCorSNPs}
snpLDcor <- snpLD %>% inner_join(snpcor, by = c("Var1", "Var2"))
```

`r nrow(snpLDcor)` pairs of SNPs with both a Pearson correlation coefficient equal to 1 and LD > 0.99 (with LD estimated as the $r$ correlation coefficient among the SNPs).


# Exploring the genomic data

This section is only done on the genomic dataset without MAF.

```{r LoadGenomicDataNOMAF}
df <- read.csv(here("data/DryadRepo/FormattedFilteredGenomicData_AlleleCounts_withoutmaf.csv"), row.names = 1) 
```


## How many SNPs genotyped in each assay?

We look at how many SNPs were genotyped in the Infinium assay, in the Affymetric assay and in both. 

```{r NumberSNPgenotypedAssays}
codes <- read.csv(here("data/GenomicData/SnpCodesMatching.csv"))

temp_df <- read.csv(here("data/GenomicData/SnpCodesMatching.csv")) %>% 
  merge(df %>% rownames_to_column(var="snp_ID"),by="snp_ID")

nb <- sapply(temp_df[,1:4], function(x) sum(!(is.na(x))))
```

Among the `r nb[[1]]` SNPs used in this paper (after filtering), **`r nb[[3]]` were genotyped in the Affymetrix assay**, **`r nb[[4]]` in the Infinium assay** and **`r temp_df[,c(1,3,4)] %>% drop_na() %>% nrow()` were genotyped by both assays**.

## How many clones genotyped in each assay?

We want to know the same for the clones, i.e. how many clones were genotyped by each assay and by both.

```{r NumberClonesGenotypedAssays}
temp_df <- data[,1:2] %>% filter(clone %in% colnames(df))

# sapply(temp_df, function(x) sum(is.na(x))) # (to check) No NAs, great .

table(temp_df$assay)
```

All clones were genotyped with both the Affymetrix and Infinium assays.

## Average and max number of missing data per clone

Frequency of missing data per clone:

```{r MissingDataPerclone}
tab <- df %>% 
  sapply(function(x) sum(is.na(x))*100/nrow(df)) %>% 
  as.data.frame() %>%
  set_colnames(c("na_freq")) %>% 
  rownames_to_column(var = "clone") %>% 
  as_tibble() %>% 
  arrange(desc(na_freq)) %>% 
  mutate(id=1:nrow(.))

# Visualization
plot(tab$id,tab$na_freq,
     pch=20,
     xlab="Clone number",
     ylab="Percentage of missing data per clone")

# Table ordered in decreasing order
# so clones with the highest frequencies of missing data are first
tab[c(1:10),1:2]  %>% kable_mydf(boldfirstcolumn = F, font_size = 11)
```

The **maximum frequency of missing data per clone** is **`r round(max(tab$na_freq),2)`%** and the **average frequency of missing data per clone** is **`r round(mean(tab$na_freq),2)`%**.

## Number of clones per population

```{r GenerateTableLatexSuppInfoNbClonesPerProv}
# We generate a table for the Supplementary Information of the paper, with the number of clones (i.e. trees) per provenance:
tibble(Provenance=str_sub(colnames(df),1,3),clon=colnames(df)) %>% 
  group_by(Provenance) %>% 
  count(name="Number of clones") %>% 
  xtable(type="latex",digits=3) %>% 
  print(file=here('tables/NbclonesPerPopulationGenomicData.tex'),include.rownames=FALSE)
```

# Imputation of missing data

We impute the missing data in both datasets; with or without minor allele frequencies.

Percentage of missing data:
  
  - in the genomic dataset with MAF: `r round(sum(is.na(list_geno$list_withmaf$df)) * 100 / (dim(list_geno$list_withmaf$df)[1]*dim(list_geno$list_withmaf$df)[2]),2)`%.
  
  - in the genomic dataset without MAF:`r round(sum(is.na(list_geno$list_withoutmaf$df)) * 100 / (dim(list_geno$list_withoutmaf$df)[1]*dim(list_geno$list_withoutmaf$df)[2]),2)`%.

We are going to impute the missing values using the most common allele at each SNP within the main gene pool of the clone. So here, the genotypes are classified according to the main gene pool they belong to (although we acknowledge that some genotypes are admixed and belong to different gene pools).

We load the dataset in which there is the proportion of belonging of each genotype to a given gene pool [from @jaramillo2015molecular].

```{r LoadingPropAncestry}
prop <- read.csv(here("data/DryadRepo/PopulationStructureCorrea2015.csv")) %>% 
  dplyr::select(clon,pop,paste0(rep("Q",6),1:6),max.Q)

# To check whether some clones in the genomic data have no population structure data.
setdiff(colnames(list_geno$list_withmaf$df),prop$clon) # this is ok :)

mainGP <- prop %>% dplyr::select(clon,max.Q) # extract the main gene pool for each clone.


list_geno <- lapply(list_geno, function(x){
 
df <-  x$df %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("clon") %>%  
  left_join(mainGP,by="clon") %>% 
  dplyr::select(clon,max.Q,everything())
  
list(df=df)
  
})

# viz geno data without MAF
list_geno$list_withoutmaf$df[1:10,1:10]  %>% kable_mydf(boldfirstcolumn = T, font_size = 11)
```

```{r CheckingImputation1}
# We can check that the imputation worked fine
# For that, before imputing, we find some SNPs with NAs
# we do it in the geno dataset without MAF

subset <- list_geno$list_withoutmaf$df[list_geno$list_withoutmaf$df$max.Q=="Q5",]
set.seed(6)
vec_snps_with_nas <- sample(colnames(subset)[colSums(is.na(subset)) > 0],10,replace=F) #extract some SNPs with NAs

subset_df <- lapply(vec_snps_with_nas, function(x){
  subset %>% 
  tabyl(any_of(x)) %>% 
  dplyr::select(1:2) %>% 
  set_colnames(c("genotype","count")) %>%  
  pivot_wider(names_from = "genotype", values_from = "count") %>%
  mutate(expected_value_after_imp=colnames(.)[which.max(across(everything(.)))],snp = x)
}) %>% 
  bind_rows() %>% 
  dplyr::rename(nb_0_bf_imp = `0`,
                nb_1_bf_imp = `1`,
                nb_2_bf_imp = `2`,
                nb_NA_bf_imp = `NA`)
```


We impute the missing data:

```{r NAsImputation}
# Replace NAs by (1) subseting by GP, (2) imputing NAs, (3) reassembling the GP.
list_geno <- mclapply(list_geno, function(x){
 
for(i in unique(x$df$max.Q)){
  subset <- x$df[x$df$max.Q==i,]
  subset <- apply(subset[,3:ncol(subset)], 2, function(x) replace(x, is.na(x), as.numeric(names(which.max(table(x))))))
  x$df[x$df$max.Q==i,3:ncol(x$df)] <- subset
}
  
list(df=x$df)
  
})
```


```{r CheckingImputation2}
# If we want to check that the imputation worked fine
  # Look at the SNPs that had NAs before imputation
subset <- list_geno$list_withoutmaf$df[list_geno$list_withoutmaf$df$max.Q=="Q5",]
subset_df_2 <- lapply(vec_snps_with_nas, function(x){
  subset %>% 
  tabyl(any_of(x)) %>% 
  dplyr::select(1:2) %>% 
  set_colnames(c("genotype","count")) %>%  
  pivot_wider(names_from = "genotype", values_from = "count") %>%
  mutate(snp = x)
}) %>% 
  bind_rows() %>% 
  dplyr::rename(nb_0_after_imp = `0`,
                nb_1_after_imp = `1`,
                nb_2_after_imp = `2`) %>% 
  left_join(subset_df, by="snp") %>% 
  dplyr::select(snp, contains("bf_imp"),expected_value_after_imp, everything())


subset_df_2  %>% kable_mydf(boldfirstcolumn = F, font_size = 11)
```

Formating the dataset (reattributing clone names, clones in columns and snp in rows, etc.):

```{r ReformatImputedDataset}
list_geno <- mclapply(list_geno, function(x){
  
x$df <- x$df %>% 
  column_to_rownames(var="clon") %>% 
  dplyr::select(-max.Q) %>% 
  t() %>% 
  as.data.frame()

list(df=x$df)

})


# Visualization:
list_geno$list_withoutmaf$df[1:10,1:10]  %>% kable_mydf(boldfirstcolumn = T, font_size = 11)
```

 
# Exporting the imputed genomic data

We export the genomic datasets imputed for missing data in the DRYAD repository. 

## Allele counts

```{r ExportImputedGenomicDataAlleleCounts}
for(i in 1:length(list_geno)){
  specMAF <- names(list_geno[i])[1] %>% str_sub(6, -1)
  
  list_geno[[i]]$df %>% 
  rownames_to_column(var="snp_ID") %>% 
  write_csv(here(paste0("data/DryadRepo/ImputedGenomicData_AlleleCounts_",specMAF,".csv")),
            append=FALSE)
}
```

##  Allele frequencies

```{r ExportImputedGenomicDataAllelefrequencies}
# Calculate allele frequencies
list_geno <- mclapply(list_geno, function(x){

list(df=compute_allele_freq(x$df))

})


# export in csv
for(i in 1:length(list_geno)){
  specMAF <- names(list_geno[i])[1] %>% str_sub(6, -1)
  
  list_geno[[i]]$df %>% 
  write_csv(here(paste0("data/DryadRepo/ImputedGenomicData_AlleleFrequencies_",specMAF,".csv")),
            append=FALSE)
}


list_geno$list_withoutmaf$df[1:10,1:8]  %>% 
  mutate(across(where(is.numeric), round, 2)) %>%
  kable_mydf(boldfirstcolumn = F, font_size = 11)
```


# Session information

```{r SessionInfo}
devtools::session_info()
```

# References