---
title: "Formatting, exploring and imputing genomic data"
author: "Juliette Archambeau"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    # code_fold: hide
    toc: true
    toc_depth: 4
    toc_float:
       collapsed: false
    number_sections: true
bibliography: references.bib
editor_options: 
  chunk_output_type: console
---

<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>

<style type="text/css">
div.main-container {
  max-width: 2000px;
  margin-left: auto;
  margin-right: auto;
}
</style>


```{css, echo=FALSE}
pre {
  max-height: 300px;
  overflow-y: auto;
}

pre[class] {
  max-height: 600px;
}
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width = 5,fig.height = 4,cache=F)
options(width = 300)
library(knitr)      # CRAN v1.26
library(dplyr)      # CRAN v1.0.0
library(tidyverse)  # CRAN v1.3.0
library(readxl)     # CRAN v1.3.1
library(magrittr)   # CRAN v2.0.1
library(kableExtra) # CRAN v1.1.0
library(genetics)   # to calculate LD
library(xtable)
library(reshape2)
library(parallel)
```


# Raw genomic data

## Loading

```{r LoadRawGenomicData}
# Load the raw genomic data:
data <- read.csv("data/DryadRepo/RawGenomicData.csv")
```

Below is a subset of the whole genomic dataset. NAs are indicated with `---`.

Meaning of the columns:

  - `clone`: clone ID.
  
  - `assay`: Assay in which the clone was genotyped, either the Infinium assay (`only_Inf`), the Axiom assay (`only_Affx`) or both assays (`both_Inf_Affx`).
  
  - `snp_1` --> `snp_14016`:  genotype for each of the 14,016 SNPs.


```{r VizRawGenomicData, echo=F}
# Visualizing the dataset:
data[40:50,1:13]  %>% 
  kable() %>%  
  kable_styling(font_size=11,
                bootstrap_options = c("stripped","hover", "condensed"), full_width = F)
```

## Formatting

For each SNP, we convert the letters (e.g. A/A, A/G) to numbers (0,1 or 2).

```{r FormattingRawGenomicDataFromLettersToNumbers}
# We apply a function to each column (i.e SNP) of the raw genomic data, 
# after remonving the first two columns, i.e. the clone ID and the assay in which the clone was genotyped

list.alleles <- lapply(colnames(data[,-c(1,2)]), function(x){ 
  
  vec <- data[,x]
  
  tab <- table(data[,x]) %>%  as.matrix() %>%  t() %>% as.data.frame() %>% dplyr::select(-contains("---"))
  
  
  if(length(colnames(tab))==1){         # monomorphic case, e.g. only A/A
    major <- names(which.max(tab))
    
    case_when(vec == major ~ 0) 
    
  } else if(length(colnames(tab))==2){  # case where there is no homozygote with the rare allele, e.g. A/A and A/G
    
    major <- names(which.max(tab))
    mid <- names(which.min(tab))
    
    case_when(vec == major ~ 0,
              vec == mid   ~ 1) 
    
  } else if (length(colnames(tab))==3){ # common case, e.g. A/A, A/G and G/G
    
    if(tab[,1]==tab[,3]){               # to account for cases where there is the same number of A/A and G/G 
      major <- colnames(tab)[1]
      minor <- colnames(tab)[3]
      } else {
    major <- names(which.max(tab[,c(1,3)]))
    minor <- names(which.min(tab[,c(1,3)]))
      }
    mid <- setdiff(names(tab),c(major,minor))  
    
    case_when(vec == major ~ 0,
              vec == minor ~ 2,
              vec == mid   ~ 1) 
  }
})

names(list.alleles) <- colnames(data[,-c(1,2)]) # reattribute the SNP ID to list names

df <- list.alleles %>% 
  as_tibble() %>% 
  t()

colnames(df) <- data$clone # attribute clone ID
```

```{r VizFormattedGenomicData, echo=F}
df[1:11,40:50]  %>% 
  kable() %>%  
  kable_styling(font_size=11,
                bootstrap_options = c("stripped","hover", "condensed"), full_width = F)
```

# Filtering genomic data

## Filtering monomorphic SNPs

```{r IdentifyingMonomorphicSNPs}
mono <- df %>% 
  t() %>% 
  as_tibble() %>% 
  sapply(function(x) sum(x,na.rm=T)) %>% 
  as.data.frame() %>%
  set_colnames(c("sum")) %>% 
  rownames_to_column(var = "snp") %>% 
  as_tibble() %>% 
  filter(sum==0)
```

`r nrow(mono)` SNPs are monomorphic and therefore, we are going to remove them.


```{r RemovingMonomorphicSNPs}
df <- df[!(row.names(df) %in% mono$snp),]
```

<span style="color: red;">There are `r nrow(df)` non-monomorphic SNPs left.</span>



## Removing the ROD population

We remove the clones from the ROD (Rodoiros) population, as they come from a plantation of material from Central Spain in the Iberian Atlantic region.

```{r RmROD}
df <- df %>% 
  as.data.frame() %>% 
  dplyr::select(-contains("ROD"))
```


## Filtering based on NAs and MAF

### Missing data per clone

```{r MissingDataPerClone, warning=F, fig.height=6,fig.width=10}
tab <- df %>% 
  as_tibble() %>% 
  sapply(function(x) sum(is.na(x))*100/nrow(df)) %>% 
  as.data.frame() %>%
  set_colnames(c("freq.NAs")) %>% 
  rownames_to_column(var = "clone") %>% 
  as_tibble() %>% 
  arrange(freq.NAs) %>% 
  mutate(id=1:nrow(.))

plot(tab$id,tab$freq.NAs,
     pch=20,
     xlab="Clone number",
     ylab="Percentage of missing data per clone")
```

`r nrow(tab[tab$freq.NAs<20,])` clones have less than 20% missing data, `r nrow(tab[tab$freq.NAs<18,])` clones have less than 18% and `r nrow(tab[tab$freq.NAs<15,])` have less than 15%.

```{r RemoveCloneMissingData}
# Specifying the chosen threshold:
threshold.NA.per.clone <- 18

# Removing clones with more than 18% missing data
df <- df %>% dplyr::select(tab[tab$freq.NAs<threshold.NA.per.clone,]$clone)
```

**<span style="color: red;">We keep the `r nrow(tab[tab$freq.NAs<threshold.NA.per.clone,])` clones that have less than `r threshold.NA.per.clone`% missing data.</span>**

### Minor allele frequencies

We want to remove *minor allele frequencies* for gene-environment association analyses, but not for population genetic structure estimation. Therefore, we produce two datasets: one with MAF and the other without, that we store in a list.

```{r MAF, warning=F, fig.height=6,fig.width=10}
tab <- df %>% 
  t() %>% 
  as_tibble() %>% 
  sapply(function(x) sum(x,na.rm=T)*100/(2*nrow(.))) %>% 
  as.data.frame() %>%
  set_colnames(c("MAF")) %>% 
  rownames_to_column(var = "snp") %>% 
  as_tibble() %>% 
  arrange(MAF) %>% 
  mutate(id=1:nrow(.))

plot(tab$id,tab$MAF,
     pch=20,
     xlab="SNP number",
     ylab="Minor allele frequencies")
```

**<span style="color: red;">`r nrow(tab[tab$MAF<1,])` SNPs have MAF < 1%.</span>** 

We remove the SNPs with MAF < 1%.

```{r RmMissingDataMAF}
list.geno <- list(list.maf=list(df = df),
                  list.nomaf=list(df = df[!(row.names(df) %in% tab[tab$MAF<1,]$snp),]))
```

**<span style="color: red;">In the dataset without MAF, there are `r nrow(list.geno$list.nomaf$df)` SNPs left.</span>** 


### Missing data per SNP

```{r MissingDataPerSNP, warning=F}
RmMissingDataPerSNP <- function(x){
  
tab <- x$df %>% 
  t() %>% 
  as_tibble() %>% 
  sapply(function(x) sum(is.na(x))*100/nrow(.)) %>% 
  as.data.frame() %>%
  set_colnames(c("freq.NAs")) %>% 
  rownames_to_column(var = "snp") %>% 
  as_tibble() %>% 
  arrange(freq.NAs) %>% 
  mutate(id=1:nrow(.))

df <- x$df[(row.names(x$df) %in% tab[tab$freq.NAs<20,]$snp),] %>% 
  dplyr::select(order(colnames(x$df)))

list(df=df,tab=tab)
}

list.geno <- lapply(list.geno,RmMissingDataPerSNP)
```


`r nrow(list.geno$list.nomaf$tab[list.geno$list.nomaf$tab$freq.NAs<20,])` SNPs have the percentage of missing data lower than 20% and `r nrow(list.geno$list.nomaf$tab[list.geno$list.nomaf$tab$freq.NAs<15,])` lower than 15%.

**<span style="color: red;">We keep SNPs with missing data lower than 20% and so there are `r nrow(list.geno$list.nomaf$df)` SNPs left.</span>**


```{r PlotMissingDataPerSNP, fig.height=6,fig.width=10}
plot(list.geno$list.nomaf$tab$id,list.geno$list.nomaf$tab$freq.NAs,
     pch=20,
     xlab="SNP number",
     ylab="Percentage of missing data per SNP",
     main="Genomic dataset without MAF")
```

# Exporting the formatted and filtered genomic datasets

##  Allele counts (i.e. 0/1/2)


```{r ExportingFormattedGenomicDataAlleleCounts}
for(i in 1:length(list.geno)){
  specMAF <- names(list.geno[i])[1] %>% str_sub(6, -1) %>% str_to_upper()
  
  list.geno[[i]]$df %>% 
  rownames_to_column(var="snp_ID") %>% 
  write_csv(paste0("data/DryadRepo/FormattedFilteredGenomicData_AlleleCounts_",
                   specMAF,"_",
                   ncol(list.geno[[i]]$df),"clones_",nrow(list.geno[[i]]$df),"snps.csv"),
            append=FALSE)
}

# Reminder
# to download the dataset with snp_ID as first column: use `read.csv` with missing row.names
# to download the dataset with snp_ID as row names: use `read.csv` with row.names = 1
```

`r sum(is.na(list.geno$list.maf$df))` missing data in the dataset with MAF, and `r sum(is.na(list.geno$list.nomaf$df))` missing data in the dataset without MAF.

##  Allele frequencies


```{r ExportingFormattedGenomicDataAllelefrequencies}
# Only useful for the dataset without MAF

CalcAlleleFreq <- function(x){
  
x %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var="pop") %>% 
  mutate(pop=str_sub(pop,1,3)) %>% 
  group_by(pop) %>% 
  summarise_all(~sum(., na.rm = TRUE)/((n()-sum(is.na(.)))*2))
  
}

df.freq <- CalcAlleleFreq(list.geno$list.nomaf$df) 

df.freq %>% 
  write_csv(paste0("data/DryadRepo/FormattedFilteredGenomicData_AlleleFrequencies_NOMAF_",
                   ncol(df.freq),"clones_",nrow(df.freq),"snps.csv"),
            append=FALSE)

df.freq[1:10,1:8] %>% 
  kable() %>%  
  kable_styling(bootstrap_options = c("stripped","hover", "condensed"), full_width = F)
```

`r sum(is.na(df.freq))` missing data in this dataset.


# Collinearity among SNPs

This part is only done on genomic data without MAF.

## Identifying duplicates

The first way to identify the duplicated SNPs is by using the function `duplicated`. 

```{r IdentifyPerfectDuplicates}
df <- read.csv("data/DryadRepo/FormattedFilteredGenomicData_AlleleCounts_NOMAF_454clones_9817snps.csv", row.names = 1) 

dup <- df[duplicated(df) | duplicated(df, fromLast=TRUE), ]
```

There are `r nrow(dup)` duplicates (=SNPs that have exactly the same genotype as at least one other SNP).

Then, we determine how many copies there are for each duplicate. 

```{r CountCopiesDuplicates}
count.duplicates <- function(DF){
  x <- do.call('paste', c(DF, sep = '\r'))
  ox <- order(x)
  rl <- rle(x[ox])
  cbind(DF[ox[cumsum(rl$lengths)],,drop=FALSE],count = rl$lengths)
  
}
countdup <- count.duplicates(df %>% as.data.frame()) 

tab <- table(countdup$count) %>% 
  as.data.frame() %>% 
  setNames(c("SnpNumberWithinGroups","GroupNumber"))
  
tab  %>% 
  kable() %>%  
  kable_styling(bootstrap_options = c("stripped","hover", "condensed"), full_width = F)
```

In this table, `r colnames(tab)[1]` is the number of duplicated SNPs within each group (ie groups of 2 SNPs, 3 SNPs, 4 SNPs, etc) and  `r colnames(tab)[2]` is the number of groups (or the number of non-duplicated SNPs for the first row). 

So it means that there are:

  - `r tab[1,2]` non-duplicated SNPs.
  - `r tab[2,2]` groups of `r tab[2,1]` duplicated SNPs.
  - `r tab[3,2]` groups of `r tab[3,1]` duplicated SNPs.
  - `r tab[4,2]` group of `r tab[4,1]` duplicated SNPs.
  - `r tab[5,2]` group of `r tab[5,1]` duplicated SNPs.
  - `r tab[6,2]` group of `r tab[6,1]` duplicated SNPs.
  - `r tab[7,2]` group of `r tab[7,1]` duplicated SNPs.

```{r ChecksNumberDuplicates, eval=F}
# checking that the numbers are coherent

# we should recover the total number of SNPs
TotNbSNPs <- tab %>% 
  dplyr::mutate(product=as.numeric(SnpNumberWithinGroups)*as.numeric(GroupNumber)) %>% 
  dplyr::summarise(sum=sum(product))
TotNbSNPs # that's ok!

# we should recover the total number of duplicated SNPs
TotNbSNPs - tab$GroupNumber[1] # that's ok too!
```

But, there are two situations not accounted for by the function `duplicated`. 

  - First, if two SNPs have exactly the same genotype for all clones, except one clone where one of the two SNPs has a value of "1" for instance, and the other SNP has the value "NA". In this case, the two SNPs will not be considered as duplicates. 
  - Second, if the two SNPs covary perfectly but one have "0" when the other have "2", and vice versa, they will not be counted as duplicates neither. For instance:
      - SNP_1: 0 0 1 1 2 2 
      - SNP_2: 2 2 1 1 0 0 

To add these situations in the count of duplicated SNPs, I selected all the pairs of SNPs with a correlation coefficient of 1. 

```{r IdentifyCorrelatedSNPs, warning=F}
# Invert the genomic matrix
tdf <- df %>% t() %>% as.data.frame() 

# Create a matrix of correlation for all SNPs (12483snps x 12483snps)
matcor <- cor(tdf, use = "pairwise.complete.obs")
matcor[1:10,1:10] %>% 
  kable() %>%  
  kable_styling(bootstrap_options = c("stripped","hover", "condensed"), full_width = F)

# Keep only the upper triangle of the matrix and remove values in the diagonal
matcor[lower.tri(matcor,diag=T)] <- NA
matcor[1:10,1:10] %>% 
  kable() %>%  
  kable_styling(bootstrap_options = c("stripped","hover", "condensed"), full_width = F)
```


<span style="color: red;">`r length(which(matcor==1))` pairs of SNPs have a correlation coefficient of 1.</span>

```{r ExtractCorrelatedSNPs}
snpcor <- matcor %>% 
  melt() %>% 
  na.omit() %>% 
  dplyr::filter(value==1) %>% 
  dplyr::select(-value)

rm(matcor) # to free memory space
```


## Pairwise LD

 We calculate the *pairwise linkage disequilibrium* among SNPs with the `genetics` package.

```{r CalculateLD,eval=F}
# format genomic data for the `LD` function of the `genetics` package
LD <- data %>% 
  dplyr::select(-clone,-assay) %>% 
  dplyr::select(row.names(df)) %>% # keep only filtered SNPs
  # !!Comment!! do not merge this line with the previous one, it does not anymore if I do it
  makeGenotypes() 

# calculate LD (very very very long)
LD <- LD(LD)

saveRDS(LD, file="outputs/LD/LDGeneticsRpackage.rds")
```

```{r IdentifyHighLDSNPs}
LDmat <- readRDS("outputs/LD/LDGeneticsRpackage.rds")
LDmat <- LDmat$`R^2`

snpLD <- LDmat %>% 
  melt() %>% 
  na.omit() %>% 
  dplyr::filter(value>0.99) %>% 
  dplyr::select(-value)

rm(LDmat) # to free memory space
```

`r nrow(snpLD)` pairs of SNPs with LD > 0.99 (with LD estimated as the $r$ correlation coefficient among the SNPs).

```{r CommonCorSNPs}
snpLDcor <- snpLD %>% inner_join(snpcor, by = c("Var1", "Var2"))
```

`r nrow(snpLDcor)` pairs of SNPs with both a Pearson correlation coefficient equal to 1 and LD > 0.99 (with LD estimated as the $r$ correlation coefficient among the SNPs).

<span style="color: red;">**Do we prune the collinear SNPs?**</span>

# Exploring the genomic data

This section is only done on the genomic dataset without MAF.

```{r LoadGenomicDataNOMAF}
df <- read.csv("data/DryadRepo/FormattedFilteredGenomicData_AlleleCounts_NOMAF_454clones_9817snps.csv", row.names = 1) 
```


## How many SNPs genotyped in each assay?

We look at how many SNPs were genotyped in the Infinium assay, in the Affymetric assay and in both. 

```{r NumberSNPgenotypedAssays}
codes <- read.csv("data/DryadRepo/SnpCodesMatching.csv")

temp.df <- read.csv("data/DryadRepo/SnpCodesMatching.csv") %>% 
  merge(df %>% rownames_to_column(var="snp_ID"),by="snp_ID")

nb <- sapply(temp.df[,1:4], function(x) sum(!(is.na(x))))
```

Among the `r nb[[1]]` SNPs used in this paper (after filtering), **`r nb[[3]]` were genotyped in the Affymetrix assay**, **`r nb[[4]]` in the Infinium assay** and **`r temp.df[,c(1,3,4)] %>% drop_na() %>% nrow()` were genotyped by both assays**.

## How many clones genotyped in each assay?

We want to know the same for the clones, i.e. how many clones were genotyped by each assay and by both.

```{r NumberClonesGenotypedAssays}
temp.df <- data[,1:2] %>% filter(clone %in% colnames(df))

# sapply(temp.df, function(x) sum(is.na(x))) # (to check) No NAs, great .

table(temp.df$assay)
```

All clones were genotyped with both the Affymetrix and Infinium assays.

## Average and max number of missing data per clone

Frequency of missing data per clone:

```{r MissingDataPerclone}
tab <- df %>% 
  sapply(function(x) sum(is.na(x))*100/nrow(df)) %>% 
  as.data.frame() %>%
  set_colnames(c("freq.NAs")) %>% 
  rownames_to_column(var = "clone") %>% 
  as_tibble() %>% 
  arrange(desc(freq.NAs)) %>% 
  mutate(id=1:nrow(.))

# Visualization
plot(tab$id,tab$freq.NAs,
     pch=20,
     xlab="Clone number",
     ylab="Percentage of missing data per clone")

# Table ordered in decreasing order
# so clones with the highest frequencies of missing data are first
tab[c(1:10),1:2] %>%
  kable() %>%  
  kable_styling(font_size=11,
                bootstrap_options = c("stripped","hover", "condensed"), full_width = F)
  
```

The **maximum frequency of missing data per clone** is **`r round(max(tab$freq.NAs),2)`%** and the **average frequency of missing data per clone** is **`r round(mean(tab$freq.NAs),2)`%**.

## Number of clones per population

```{r GenerateTableLatexSuppInfoNbClonesPerProv}
# We generate a table for the Supplementary Information of the paper, with the number of clones (i.e. trees) per provenance:
tibble(Provenance=str_sub(colnames(df),1,3),clon=colnames(df)) %>% 
  group_by(Provenance) %>% 
  count(name="Nb.clones") %>% 
  xtable(type="latex",digits=3) %>% 
  print(file='tables/NbclonesPerPopulationGenomicData.tex',include.rownames=FALSE)
```

# Imputation of missing data

We will impute the missing data in both datasets with or without minor allele frequencies.

Percentage of missing data:
  
  - in the genomic dataset with MAF: `r round(sum(is.na(list.geno$list.maf$df)) * 100 / (dim(list.geno$list.maf$df)[1]*dim(list.geno$list.maf$df)[2]),2)`%.
  
  - in the genomic dataset without MAF:`r round(sum(is.na(list.geno$list.nomaf$df)) * 100 / (dim(list.geno$list.nomaf$df)[1]*dim(list.geno$list.nomaf$df)[2]),2)`%.

We are going to impute the missing values using the most common allele at each SNP within the main gene pool of the clone. So here, the genotypes are classified according to the main gene pool they belong to (although we acknowledge that some genotypes are admixed and belong to different gene pools).

We load the dataset in which there is the proportion of belonging of each genotype to a given gene pool [from @jaramillo2015molecular].

```{r LoadingPropAncestry}
prop <- read.csv("data/DryadRepo/PopulationStructureCorrea2015.csv") %>% 
  dplyr::select(clon,prov,paste0(rep("Q",6),1:6),max.Q)

# To check whether some clones in the genomic data have no population structure data.
setdiff(colnames(list.geno$list.maf$df),prop$clon) # this is ok :)

mainGP <- prop %>% dplyr::select(clon,max.Q) # extract the main gene pool for each clone.


list.geno <- lapply(list.geno, function(x){
 
df <-  x$df %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("clon") %>%  
  left_join(mainGP,by="clon") %>% 
  dplyr::select(clon,max.Q,everything())
  
list(df=df)
  
})

# viz geno data without MAF
list.geno$list.nomaf$df[1:10,1:10] %>%
  kable() %>%
  kable_styling(bootstrap_options = c("hover", "condensed"), full_width = F) %>%
  column_spec(1, bold = T)
```

```{r CheckingImputation1}
# We can check that the imputation worked fine
# For that, before imputing, we find some SNPs with NAs
# we do it in the geno dataset without MAF

subset <- list.geno$list.nomaf$df[list.geno$list.nomaf$df$max.Q=="Q5",]
sample(colnames(subset)[colSums(is.na(subset)) > 0],3,replace=F) #extract some SNPs with NAs
sum(is.na(subset)) # 14,902 NAs in Q5 gene pool

table(subset[,"snp_12238"]) # NAs of this SNP should take the value '2' after imputation.
table(subset[,"snp_10182"]) # NAs of this SNP should take the value '1' after imputation.
table(subset[,"snp_7628"])  # NAs of this SNP should take the value '2' after imputation.
```


We impute the missing data:

```{r NAsImputation}
# Replace NAs by (1) subseting by GP, (2) imputing NAs, (3) reassembling the GP.
list.geno <- mclapply(list.geno, function(x){
 
for(i in unique(x$df$max.Q)){
  subset <- x$df[x$df$max.Q==i,]
  subset <- apply(subset[,3:ncol(subset)], 2, function(x) replace(x, is.na(x), as.numeric(names(which.max(table(x))))))
  x$df[x$df$max.Q==i,3:ncol(x$df)] <- subset
}
  
list(df=x$df)
  
})
```


```{r CheckingImputation2}
# If we want to check that the imputation worked fine
  # Look at the SNPs that had NAs before imputation
subset <- list.geno$list.nomaf$df[list.geno$list.nomaf$df$max.Q=="Q5",]
sum(is.na(subset)) # 0 NAs now!

table(subset[,"snp_12238"]) # great! (NAs of this SNP has taken the values '2' after imputation)
table(subset[,"snp_10182"]) # great! (The missing value of this SNP has taken the value '1' after imputation)
table(subset[,"snp_7628"])  # great! (The missing value of this SNP has taken the value '2' after imputation)
```

Formating the dataset (reattributing clone names, clones in columns and snp in rows, etc.):

```{r ReformatImputedDataset}
list.geno <- mclapply(list.geno, function(x){
  
x$df <- x$df %>% 
  column_to_rownames(var="clon") %>% 
  dplyr::select(-max.Q) %>% 
  t() %>% 
  as.data.frame()

list(df=x$df)

})


# Visualization:
list.geno$list.nomaf$df[1:10,1:10] %>%
  kable() %>%
  kable_styling(bootstrap_options = c("hover", "condensed"), full_width = F) %>%
  column_spec(1, bold = T)
```

 
# Exporting the imputed genomic data


## Allele counts

```{r ExportImputedGenomicDataAlleleCounts}
for(i in 1:length(list.geno)){
  specMAF <- names(list.geno[i])[1] %>% str_sub(6, -1) %>% str_to_upper()
  
  list.geno[[i]]$df %>% 
  rownames_to_column(var="snp_ID") %>% 
  write_csv(paste0("data/DryadRepo/ImputedGenomicData_AlleleCounts_",
                   specMAF,"_",
                   ncol(list.geno[[i]]$df),"clones_",nrow(list.geno[[i]]$df),"snps.csv"),
            append=FALSE)
}
```

##  Allele frequencies

```{r ExportImputedGenomicDataAllelefrequencies}
# Calculate allele frequencies
list.geno <- mclapply(list.geno, function(x){

list(df=CalcAlleleFreq(x$df))

})


# export in csv
for(i in 1:length(list.geno)){
  specMAF <- names(list.geno[i])[1] %>% str_sub(6, -1) %>% str_to_upper()
  
  list.geno[[i]]$df %>% 
  write_csv(paste0("data/DryadRepo/ImputedGenomicData_AlleleFrequencies_",
                   specMAF,"_",
                   nrow(list.geno[[i]]$df),"pops_",ncol(list.geno[[i]]$df[,-1]),"snps.csv"),
            append=FALSE)
}



list.geno$list.nomaf$df[1:10,1:8] %>% 
  kable() %>%  
  kable_styling(bootstrap_options = c("stripped","hover", "condensed"), full_width = F)
```


# References